WorkflowSwiftUIExperimental/Sources/EnvironmentValues+ViewEnvironment.swift

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2023 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import SwiftUI
+import WorkflowUI
+
+private struct ViewEnvironmentKey: EnvironmentKey {
+    static let defaultValue: ViewEnvironment = .empty
+}
+
+public extension EnvironmentValues {
+    var viewEnvironment: ViewEnvironment {
```

Comment:
Yes, I expect that if a view's body accesses any part of `viewEnvironment`, that body will be unnecessarily reevaluated for any change to any other part of the `ViewEnvironment`. It's also cumbersome that the View has to write out e.g. `@Environment(\.viewEnvironment.marketStylesheet)`.

I think the next step is to explore mapping ViewEnvironment keys to individual SwiftUI Environment keys, perhaps using [UITraitBridgedEnvironmentKey](https://developer.apple.com/documentation/SwiftUI/UITraitBridgedEnvironmentKey) as @n8chur has suggested.

Reply:
Rob and I did some experimentation on the `@Environment` property wrapper and found that an `EnvironmentKey` observing view will not invalidate if the subset of the value your property wrapper points to (via `KeyPath`) is unchanged (using the same magic comparison that SwiftUI uses for determining changes in other systems.

For example, if you have an Environment value with two parameters in an Environment key:
```swift
struct MyThing {
    var foo = 0
    var bar = 0
}
struct MyKey: EnvironmentKey {
    static let defaultValue = MyThing()
}
extension EnvironmentValues {
    var myThing: MyThing {
        get { self[MyKey.self] }
        set { self[MyKey.self] = newValue }
    }
}
```
and you have a view that observes `\.myThing.bar`:
```swift
struct MyObservingView: View {
    @Environment(\.myThing.bar) var value
    var body: some View {
        let _ = Self._printChanges()
        Text("Bar: \(String(describing: value))")
    }
}
```
contained within another view that sets `\.myValue.foo`:
```swift
struct MyContainerView: View {
    @State var foo = 0
    var body: some View {
        VStack {
            HStack {
                Stepper("Foo", value: $foo)
                Text("\(String(describing: foo))")
            }
            MyObservingView()
        }
        .environment(\.myThing.foo, foo)
    }
}
```
Changes to `\.myValue.foo` do not case the body of views observing the `\.myValue.bar` key=path (`MyObservingView`) to be re-evaluated.

<details>
  <summary>Full code for playground</summary>

```swift
import SwiftUI


struct MyThing {
    var foo = 0
    var bar = 0
}


struct MyKey: EnvironmentKey {
    static let defaultValue = MyThing()
}


extension EnvironmentValues {
    var myThing: MyThing {
        get { self[MyKey.self] }
        set { self[MyKey.self] = newValue }
    }
}


struct MyObservingView: View {
    @Environment(\.myThing.bar) var value
    var body: some View {
        let _ = Self._printChanges()
        Text("Bar: \(String(describing: value))")
    }
}

struct MyContainerView: View {
    @State var foo = 0
    var body: some View {
        VStack {
            HStack {
                Stepper("Foo", value: $foo)
                Text("\(String(describing: foo))")
            }
            MyObservingView()
        }
        .environment(\.myThing.foo, foo)
    }
}


import PlaygroundSupport


PlaygroundPage.current.setLiveView(MyContainerView())
```
</details>

This behavior makes me feel much better about the idea of using the shape proposed here for just composing the environment, but I'm sure there are other benefits to going with the bridged key approach still. This is, for example, still very easy to naively access in a way that would be non performant:
```swift
@Environment(\.viewEnvironment) var viewEnvironment
// ...
// I think in some cases this type of access would cause body invalidations if _anything_ on the ViewEnvironment changed?
.someModifier(value: viewEnvironment.some.long.keyPath)

// instead of:
// This should only invalidate the body if the value at the keyPath changes
@Environment(\.viewEnvironment.some.long.keyPath) var value
```

If the bridging key approach doesn't work out, perhaps an incremental improvement over what we have now would be to just not expose the `viewEnvironment` key directly and instead require access via another property wrapper which wraps it (e.g. `@ViewEnvironment`)?

Reply:
E.g. we could hide the `viewEnvironmnt` `EnvironmentValues` behind a property wrapper / view modifier set like this:
```swift
@propertyWrapper
public struct ViewEnvironmentReader<Value>: DynamicProperty {
    @Environment
    public var wrappedValue: Value

    public init(_ keyPath: KeyPath<ViewEnvironment, Value>) {
        _wrappedValue = Environment((\EnvironmentValues.viewEnvironment).appending(path: keyPath))
    }
}

extension View {

    public func viewEnvironment<Value>(_ keyPath: WritableKeyPath<ViewEnvironment, Value>, _ value: Value) -> some View {
        environment(
            (\EnvironmentValues.viewEnvironment).appending(path: keyPath),
            value
        )
    }
}
```